src/Package/Impl/Search/VsSearchTask.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.R.Components.Search;
+using Microsoft.VisualStudio.Shell.Interop;
+using static Microsoft.VisualStudio.VSConstants.VsSearchTaskStatus;
+
+namespace Microsoft.VisualStudio.R.Package.Search {
+    internal class VsSearchTask : IVsSearchTask {
+        private readonly IVsSearchCallback _searchCallback;
+        private readonly ISearchHandler _handler;
+        private readonly CancellationTokenSource _cts;
+        private long _taskStatus;
+
+        public uint Id { get; }
+        public int ErrorCode { get; }
+        public IVsSearchQuery SearchQuery { get; }
+        public uint Status => (uint)_taskStatus;
+
+        public VsSearchTask(uint dwCookie, IVsSearchQuery pSearchQuery, IVsSearchCallback pSearchCallback, ISearchHandler handler, CancellationTokenSource cts) {
+            if (pSearchQuery == null) {
+                throw new ArgumentNullException(nameof(pSearchQuery));
+            }
+
+            if (pSearchCallback == null) {
+                throw new ArgumentNullException(nameof(pSearchCallback));
+            }
+
+            if (dwCookie == VSConstants.VSCOOKIE_NIL) {
+                throw new ArgumentException(nameof(dwCookie));
+            }
+
+            if (string.IsNullOrEmpty(pSearchQuery.SearchString)) {
+                throw new ArgumentException(nameof(pSearchQuery));
+            }
+
+            Id = dwCookie;
+            SearchQuery = pSearchQuery;
+            ErrorCode = 0;
+
+            _searchCallback = pSearchCallback;
+            _handler = handler;
+            _cts = cts;
+            _taskStatus = (long)Created;
+        }
+
+        public void Start() {
+            if (!SetTaskStatus(Started)) {
+                return;
+            }
+
+            _handler.Search(SearchQuery.SearchString, _cts.Token).ContinueWith(SearchCompleted);
+        }
+
+        public void Stop() {
+            if (!SetTaskStatus(Stopped)) {
+                return;
+            }
+
+            _cts.Cancel();
+        }
+
+        private void SearchCompleted(Task<int> task) {
+            switch (task.Status) {
+                case TaskStatus.RanToCompletion:
+                    _searchCallback.ReportComplete(this, (uint)task.Result);
+                    break;
+                case TaskStatus.Faulted:
+                case TaskStatus.Canceled:
+                    _searchCallback.ReportComplete(this, 0);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
```

Comment:
In private method if this hits it appears to be our own bug... Perhaps assert is more appropriate?